### PR TITLE
fix: improved type checking & typescript example

### PR DESCRIPTION
**Pull request by fnknzzz**
1. More accurate type

1. Fix latest redux typing check error in typescript example

1. Add more type constrain in typescript example

related for issue #656

### CODE REVIEW

**Comment by semoal** [comment]
Thanks for your work mate.

Probably this is duplicated with:
https://github.com/rematch/rematch/pull/627
https://github.com/rematch/rematch/pull/647

We should see which one we need.

**Comment by fnknzzz** [comment]
> Thanks for your work mate.
> 
> Probably this is duplicated with:
> #627
> #647
> 
> We should see which one we need.

OK, we can test more tyepscript case

**Comment by mgambati** [comment]
Any progress on review?

File: examples/ts/count/src/store.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
+import { init, RematchRootState, RematchDispatch } from '@rematch/core'
+import models from './models'
 
 export const store = init({
 	models,
 })
 
 export type Store = typeof store
-export type Dispatch = typeof store.dispatch
+export type Dispatch = RematchDispatch<typeof models>
```
Comment (semoal) [inline examples/ts/count/src/store.ts]: Is it required to create the dispatch method like this?

Comment (fnknzzz) [reply examples/ts/count/src/store.ts]: No. This is  used in Model defination like `effects: (dispatch: Dispatch) => {}` to avoid recursive check.

**Comment by ShMcK** [comment]
Looks good, sorry haven't been keeping up with PR requests. 

I'll release changes in a batch as 1.2.0.

**Pull request merged**